### DISCUSSION TITLE
Hooking up afterViewerChanged/beforeViewerChanged

### DISCUSSION THREAD

**Discussion by lob743** [TerriaJS/terriajs#6186] [Q&A] [answered]
Hello,

Previously, in v7, we used the events `afterViewerChanged`/`beforeViewerChanged` a great deal to reload stuff upon swapping viewer mode. Now, in v8, it appears these are not hooked up, in the sense that events are never raised. Is there a reason why this was dropped after the transition to v8? 

In my attempt to hook these up, I encountered a problem.

In `TerriaViewer.ts`, I raised the `beforeViewerChanged` event inside of `destroyCurrentViewer()`, right where it says "Destroying viewer:". This works fine. However, when attempting to raise `afterViewerChanged` inside of `currentViewer`, I get an error for "Cycle detected in computation". I cannot seem to understand why one works and not the other. I would greatly appreciate any insight! 
Thanks!

**Comment by zoran995** [comment, marked as answer]
Already reported in #4752. You probably can use mobx observables to get out around that. I used a combination of `reaction` and `when` (not sure why both were needed, long time since it's implemented) to create manual subscribtion to viewer events and it seems to work fine.
```ts
  constructor() {
    reaction(
      () => this.cesiumViewer,
      () => this.addUpdateSubscription(),
      {
        equals: (a, b) => {
          return isDefined(b);
        }
      }
    );

    reaction(
      () => this.leafletViewer,
      () => this.addUpdateSubscription(),
      {
        equals: (a, b) => {
          return isDefined(b);
        }
      }
    );

    when(
      () => isDefined(this.cesiumViewer),
      () => this.addUpdateSubscription()
    );
    when(
      () => isDefined(this.leafletViewer),
      () => this.addUpdateSubscription()
    );
  }
```
```ts
  @computed
  get cesiumViewer() {
    return this.terria.cesium;
  }

  @computed
  get leafletViewer() {
    return this.terria.leaflet;
  }
```

Inside `addUpdateSubscription` there is a check which viewer is currently active and perform side effects. But it should be possible to separate them completely.

**Reply by lob743** [reply]
This solution worked wonders as a work-around!
I see that you created a PR on this matter with a more permanent solution (#6190).
Thanks a bunch for solving this @zoran995, very helpful!
